### DISCUSSION TITLE
Robot is LOST

### DISCUSSION THREAD

**Discussion by Xindictus** [gstavrinos/ai_msc_robotic_systems_assignments#46] [Q&A] [answered]
The image in `rviz2` shows something completely different from the map. I assume that means that the robot is completely lost. In such cases, how can we find out where are we? Since ground truth is a cheat and not a real world solution, how do I tell my robot "you're there"? The `TF` tree is also completely different from where the camera sees.

Soundtrack: https://youtu.be/db3B3ItcsQk?si=itSkfcqVgDbI_pA8&t=214
Adaptation: https://www.youtube.com/watch?v=KTu8iDynwNc

**Comment by gstavrinos** [comment]
lol!

Think about it in a real world scenario. If the robot was lost and you could fix its localization from code, probably it wouldn't be lost in the first place! Thus, in order to fix localization problems you generally need human intervention. Setting the initial pose from rviz is the easiest way.

For this exercise the robot shouldn't be getting lost though. Based on #45 , I assume that there is something wrong with the simulation or navigation software. 

Do you get any errors?
Is the robot moving at all when you request a new goal?
Can you post a screenshot from rviz?

**Reply by Xindictus** [reply]
Essentially, I try to move the robot to one of the table's sides, it stumbles upon the table and then starts trying to move to what it thinks are the rest of the sides, when really, it has moved to the other end of the room.

The path I think I see in `rviz2` seems correct to me, but, since the robot fails to reach even one side, I have my doubts. Attaching some `rviz2` snaps:

_What seems to be an ok path towards our goal?!_
![image](https://github.com/gstavrinos/ai_msc_robotic_systems_assignments/assets/7975922/f23b678c-7f50-4059-9c77-8464689bac85)
![image](https://github.com/gstavrinos/ai_msc_robotic_systems_assignments/assets/7975922/acde7d81-f57a-42a0-a1e8-dbcb914cef81)

_After awhile, the robot thinks (according to the `TF` tree) that it's close to the table, but, looking at the image feedback, it's completely elsewhere:_
![image](https://github.com/gstavrinos/ai_msc_robotic_systems_assignments/assets/7975922/74587fd2-ef13-489e-856c-632b621aaf9a)

The reason I started https://github.com/gstavrinos/ai_msc_robotic_systems_assignments/discussions/45 was because I thought I should add the table legs to the costmap in order for the robot to better avoid hitting the table.

**Reply by gstavrinos** [reply, marked as answer]
See #45, you appear to not have `easy_mode` enabled. If you had to solve the problem without `easy_mode`, in order to allow for the local_costmap to observe obstacles in higher planes (like the table), you should mess with the navigation configuration, and more specifically change input sensors for the obstacle layer.

**Reply by Xindictus** [reply]
**The robot is no longer lost. It navigated around the ping pong table smoothly, like butter.**
![b7832994221a14810be8b362f7f315b5](https://github.com/gstavrinos/ai_msc_robotic_systems_assignments/assets/7975922/4084bbcf-705e-47a9-b9ca-19c9392c8bd6)

---

Suppose now that we want to navigate without `easy_mode`. You listed the following required actions:
1. mess with the navigation configuration
   - Does that mean we need to mess with the `BT` tree and follow a similar approach to the following?
     - https://docs.nav2.org/behavior_trees/trees/nav_to_pose_and_pause_near_goal_obstacle.html
       e.g. set up back up distance after reaching goal pose?
       ```xml
       <BackUp backup_dist="0.50" backup_speed="0.05"/>
       ```
   - should we smooth the path ourselves using the `Simple Commander API`?
      ```python
      smoothPath(path, smoother_id=’‘, max_duration=2.0, check_for_collision=False)
      ```
   - should we dynamically update local costmap? e.g. setting a cost such as 255 if we detect a collision or predefine the obstacles so that we avoid them altogether 
2. change input sensors for the obstacle layer
   - does that mean we need to use the following types of messages?
     - `sensor_msgs/LaserScan`
     - `sensor_msgs/Range`

I've also noticed that with the `easy_mode`, the table is seen as a big rectangle on the map. Is that created using the aforementioned actions?

**Reply by gstavrinos** [reply]
1. It is much simpler. You just need to mess with the navigation configuration of the robot and more specifically the way costmap are being generated. The configuration files are located at the default/system installation of ros. Check #42 for more info on that.
2. This is almost exclusively the only thing you have to do to autonomously navigate around the table. The problem is that the sensors you listed are not 3D. This most probably wouldn't solve your problem, since the table is located in a higher plane. You would need a 3D input for that. We have previously discussed about the sensors of our robot. Maybe you could use its 3D sensor to detect the table, in order to approach the obstacle generation of the easy mode.

**Reply by Xindictus** [reply]
so we would need a `PointCloud` message to achieve that. Doesn't that mean that we need to have at least 3 (X) number of nodes, each gathering obstacle data and publishing them to a topic let's say which will orchestrate the whole collision avoidance?

**Reply by gstavrinos** [reply]
I am not really sure what you mean, but assigning new sensor inputs in the configuration file is the only requirement from the developer and the rest is solely taken care of by the navigation stack.
